### PR TITLE
feat(mpc): calculate mpc predicted trajectory in the world coordinate to improve accuracy

### PR DESCRIPTION
**Pull request by TakaHoribe**
## Description

This is a reopening of the old PR https://github.com/autowarefoundation/autoware.universe/pull/3621

In this PR, we have revised the calculation of the `predicted_trajectory`, which was previously computed using relative coordinates, to now be computed in the world coordinate.

The motivation for this change stems from the fact that the current calculation does not work in some cases. As shown in this video, particularly at the endpoints of a path, calculations become impossible.

**[Before]**

https://github.com/autowarefoundation/autoware.universe/assets/21360593/f8b1901e-198e-4dc6-a91b-01934da4d274

Moreover, in the relative coordinate system, errors from linearization and other factors were not considered, potentially leading to results that diverge from actual behavior. In this regard, the accuracy of the predicted trajectory calculations has also improved.

With this PR, the `predicted_trajectory` is improved as below:

**[After]**

https://github.com/autowarefoundation/autoware.universe/assets/21360593/d0c4c68f-9c1d-43ce-b531-1b84f6c008ba

**Important note**: the trajectory calculation should be defined based on the vehicle model, therefore the implementation is moved from `mpc.cpp` to `vehicle_model_***.cpp`.

## Related links

None

## Tests performed

Run psim, shown in the video above.

## Notes for reviewers

None

## Interface changes

None

## Effects on system behavior

The accuracy of the `predicted_trajectory` will be changed.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### Summary by CodeRabbit

- New Feature: Added `calculatePredictedTrajectory()` function to the `MPC` class for improved accuracy in predicting trajectories.
- Refactor: Updated `DynamicsBicycleModel::calculatePredictedTrajectory()` to perform trajectory calculations in world coordinates instead of relative coordinates.
- New Feature: Added `calculatePredictedTrajectory()` function to the `KinematicsBicycleModel` class for calculating predicted trajectories in world coordinates.
- New Feature: Added `calculatePredictedTrajectory()` function to the `KinematicsBicycleModelNoDelay` class for calculating predicted trajectories in world coordinates.
- Refactor: Modified `VehicleModelInterface` class by adding `const` qualifiers to getter functions for dimensions and wheelbase.

### CODE REVIEW

File: control/mpc_lateral_controller/include/mpc_lateral_controller/mpc_utils.hpp [hunk 1]
```diff
@@ -92,6 +92,13 @@ Trajectory convertToAutowareTrajectory(const MPCTrajectory & input);
  */
 void calcMPCTrajectoryArcLength(const MPCTrajectory & trajectory, std::vector<double> & arc_length);
 
+/**
+ * @brief calculate the arc length of the given trajectory
+ * @param [in] trajectory trajectory for which to calculate the arc length
+ * @return total arc length
+ */
+double calcMPCTrajectoryArcLength(const MPCTrajectory & trajectory);
```
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply control/mpc_lateral_controller/include/mpc_lateral_controller/mpc_utils.hpp:100]: The old one is to calculate arc_length at each point with `std::vector<double>`, but the new one is to calculate the arclength for the whole with `double`.

File: control/mpc_lateral_controller/include/mpc_lateral_controller/mpc_utils.hpp [hunk 2]
```diff
@@ -194,6 +201,14 @@
 void extendTrajectoryInYawDirection(
   const double yaw, const double interval, const bool is_forward_shift, MPCTrajectory & traj);
 
+/**
+ * @brief clip trajectory size by length
+ * @param [in] trajectory original trajectory
+ * @param [in] length clip length
+ * @return clipped trajectory
+ */
+MPCTrajectory clipTrajectoryByLength(const MPCTrajectory & trajectory, const double length);
```
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply control/mpc_lateral_controller/include/mpc_lateral_controller/mpc_utils.hpp:210]: No, it is implemented

File: control/mpc_lateral_controller/include/mpc_lateral_controller/vehicle_model/vehicle_model_bicycle_dynamics.hpp [hunk 3]
```diff
@@ -103,6 +103,12 @@ class DynamicsBicycleModel : public VehicleModelInterface
 
   std::string modelName() override { return "dynamics"; };
 
+  MPCTrajectory calculatePredictedTrajectory(
+    [[maybe_unused]] const Eigen::MatrixXd & a_d, [[maybe_unused]] const Eigen::MatrixXd & b_d,
+    [[maybe_unused]] const Eigen::MatrixXd & c_d, [[maybe_unused]] const Eigen::MatrixXd & w_d,
+    const Eigen::MatrixXd & x0, const Eigen::MatrixXd & Uex,
+    const MPCTrajectory & reference_trajectory, [[maybe_unused]] const double dt) const override;
```
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply control/mpc_lateral_controller/include/mpc_lateral_controller/vehicle_model/vehicle_model_bicycle_dynamics.hpp:110]: In some model, they are used now.

File: control/mpc_lateral_controller/include/mpc_lateral_controller/vehicle_model/vehicle_model_bicycle_kinematics.hpp [hunk 4]
```diff
@@ -91,6 +91,12 @@ class KinematicsBicycleModel : public VehicleModelInterface
 
   std::string modelName() override { return "kinematics"; };
 
+  MPCTrajectory calculatePredictedTrajectory(
+    [[maybe_unused]] const Eigen::MatrixXd & a_d, [[maybe_unused]] const Eigen::MatrixXd & b_d,
+    [[maybe_unused]] const Eigen::MatrixXd & c_d, [[maybe_unused]] const Eigen::MatrixXd & w_d,
+    const Eigen::MatrixXd & x0, const Eigen::MatrixXd & Uex,
+    const MPCTrajectory & reference_trajectory, [[maybe_unused]] const double dt) const override;
```
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply control/mpc_lateral_controller/include/mpc_lateral_controller/vehicle_model/vehicle_model_bicycle_kinematics.hpp:98]: implemented in cpp

File: control/mpc_lateral_controller/include/mpc_lateral_controller/vehicle_model/vehicle_model_bicycle_kinematics_no_delay.hpp [hunk 5]
```diff
@@ -87,6 +87,12 @@ class KinematicsBicycleModelNoDelay : public VehicleModelInterface
 
   std::string modelName() override { return "kinematics_no_delay"; };
 
+  MPCTrajectory calculatePredictedTrajectory(
+    [[maybe_unused]] const Eigen::MatrixXd & a_d, [[maybe_unused]] const Eigen::MatrixXd & b_d,
+    [[maybe_unused]] const Eigen::MatrixXd & c_d, [[maybe_unused]] const Eigen::MatrixXd & w_d,
+    const Eigen::MatrixXd & x0, const Eigen::MatrixXd & Uex,
+    const MPCTrajectory & reference_trajectory, [[maybe_unused]] const double dt) const override;
```
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply control/mpc_lateral_controller/include/mpc_lateral_controller/vehicle_model/vehicle_model_bicycle_kinematics_no_delay.hpp:94]: It is used in some models.

File: control/mpc_lateral_controller/src/mpc.cpp [hunk 6]
```diff
@@ -97,9 +97,9 @@
   m_raw_steer_cmd_pprev = m_raw_steer_cmd_prev;
   m_raw_steer_cmd_prev = Uex(0);
 
-  // calculate predicted trajectory
+  /* calculate predicted trajectory */
   predicted_trajectory =
-    calcPredictedTrajectory(mpc_resampled_ref_trajectory, mpc_matrix, x0_delayed, Uex);
+    calculatePredictedTrajectory(mpc_matrix, x0, Uex, mpc_resampled_ref_trajectory, prediction_dt);
```
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply control/mpc_lateral_controller/src/mpc.cpp:102]: yes

File: control/mpc_lateral_controller/src/mpc.cpp [hunk 7]
```diff
@@ -794,6 +770,23 @@
...
+    dt);
+
+  // do not over the reference trajectory
+  const auto predicted_length = MPCUtils::calcMPCTrajectoryArcLength(reference_trajectory);
+  const auto clipped_trajectory =
+    MPCUtils::clipTrajectoryByLength(predicted_mpc_trajectory, predicted_length);
+
+  const auto predicted_trajectory = MPCUtils::convertToAutowareTrajectory(clipped_trajectory);
+  return predicted_trajectory;
+}
```
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply control/mpc_lateral_controller/src/mpc.cpp:788]: yes

File: control/mpc_lateral_controller/include/mpc_lateral_controller/vehicle_model/vehicle_model_bicycle_kinematics_no_delay.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply control/mpc_lateral_controller/include/mpc_lateral_controller/vehicle_model/vehicle_model_bicycle_kinematics_no_delay.hpp:94]: Thank you. I removed unused in the declaration.

File: control/mpc_lateral_controller/src/vehicle_model/vehicle_model_bicycle_dynamics.cpp [hunk 8]
```diff
@@ -86,4 +86,37 @@ void DynamicsBicycleModel::calculateReferenceInput(Eigen::MatrixXd & u_ref)
     m_lr * m_mass / (2 * m_cf * m_wheelbase) - m_lf * m_mass / (2 * m_cr * m_wheelbase);
   u_ref(0, 0) = m_wheelbase * m_curvature + Kv * vel * vel * m_curvature;
 }
+
+MPCTrajectory DynamicsBicycleModel::calculatePredictedTrajectory(
+  [[maybe_unused]] const Eigen::MatrixXd & a_d, [[maybe_unused]] const Eigen::MatrixXd & b_d,
+  [[maybe_unused]] const Eigen::MatrixXd & c_d, [[maybe_unused]] const Eigen::MatrixXd & w_d,
+  const Eigen::MatrixXd & x0, const Eigen::MatrixXd & Uex,
+  const MPCTrajectory & reference_trajectory, [[maybe_unused]] const double dt) const
```
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply control/mpc_lateral_controller/src/vehicle_model/vehicle_model_bicycle_dynamics.cpp]: Thank you. I removed unused in the declaration.

File: control/mpc_lateral_controller/src/vehicle_model/vehicle_model_bicycle_kinematics.cpp [hunk 9]
```diff
@@ -68,4 +68,60 @@ void KinematicsBicycleModel::calculateReferenceInput(Eigen::MatrixXd & u_ref)
...
+  const auto DIM_U = getDimU();
+
+  for (size_t i = 0; i < reference_trajectory.size(); ++i) {
+    state_w = updateState(state_w, Uex.block(i * DIM_U, 0, DIM_U, 1), dt, t.vx.at(i));
+    mpc_predicted_trajectory.push_back(
+      state_w(0), state_w(1), t.z.at(i), state_w(2), t.vx.at(i), t.k.at(i), t.smooth_k.at(i),
+      t.relative_time.at(i));
+  }
+  return mpc_predicted_trajectory;
+}
```
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply control/mpc_lateral_controller/src/vehicle_model/vehicle_model_bicycle_kinematics.cpp]: thank you.

File: control/mpc_lateral_controller/src/vehicle_model/vehicle_model_bicycle_kinematics_no_delay.cpp [hunk 10]
```diff
@@ -59,4 +59,57 @@ void KinematicsBicycleModelNoDelay::calculateReferenceInput(Eigen::MatrixXd & u_
 {
   u_ref(0, 0) = std::atan(m_wheelbase * m_curvature);
 }
+
+MPCTrajectory KinematicsBicycleModelNoDelay::calculatePredictedTrajectory(
+  [[maybe_unused]] const Eigen::MatrixXd & a_d, [[maybe_unused]] const Eigen::MatrixXd & b_d,
+  [[maybe_unused]] const Eigen::MatrixXd & c_d, [[maybe_unused]] const Eigen::MatrixXd & w_d,
+  const Eigen::MatrixXd & x0, const Eigen::MatrixXd & Uex,
```
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply control/mpc_lateral_controller/src/vehicle_model/vehicle_model_bicycle_kinematics_no_delay.cpp:66]: correct

**Review by takayuki5168 — APPROVED** [review]

**Pull request merged**